pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +227,274 @@ pub mod pallet {
+
+			for asset_id in pool_details.bonded_currencies {
+				if T::Fungibles::asset_exists(asset_id.clone()) {
+					// This would fail with an LiveAsset error if there are any accounts left on any currency
+					T::Fungibles::finish_destroy(asset_id)?;
+				}
+			}
+
+			let pool_account = pool_id.clone().into();
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			if !total_collateral_issuance.is_zero() {
+				T::CollateralCurrency::transfer(
+					T::CollateralAssetId::get(),
+					&pool_account,
+					&pool_details.manager,
+					total_collateral_issuance,
+					Preservation::Expendable,
+				)?;
+			}
+
+			Pools::<T>::remove(&pool_id);
+
+			// TODO: refund deposit
+
+			Self::deposit_event(Event::Destroyed { id: pool_id });
+
+			Ok(())
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		fn do_start_refund(
+			pool_id: T::PoolId,
+			maybe_check_manager: Option<&AccountIdOf<T>>,
+			max_currencies: u32,
+		) -> Result<u32, DispatchError> {
+			let mut pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(
+				pool_details.bonded_currencies.len() <= max_currencies.saturated_into(),
+				Error::<T>::TooManyCurrencies
+			);
+
+			// refunding can only be triggered on a live pool
+			ensure!(pool_details.state.is_live(), Error::<T>::PoolNotLive);
+
+			if let Some(caller) = maybe_check_manager {
+				ensure!(pool_details.is_manager(caller), Error::<T>::Unauthorized);
+			}
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_id.clone().into());
+			// nothing to distribute
+			ensure!(!total_collateral_issuance.is_zero(), Error::<T>::NothingToRefund);
+
+			let has_holders = pool_details
+				.bonded_currencies
+				.iter()
+				.any(|asset_id| !T::Fungibles::total_issuance(asset_id.clone()).is_zero());
+			// no token holders to refund
+			ensure!(has_holders, Error::<T>::NothingToRefund);
+
+			// move pool state to refunding
+			pool_details.state.start_refund();
+			Pools::<T>::set(&pool_id, Some(pool_details.clone()));
+
+			Self::deposit_event(Event::RefundingStarted { id: pool_id });
+
+			Ok(pool_details.bonded_currencies.len().saturated_into())
+		}
+
+		fn do_start_destroy_pool(
+			pool_id: T::PoolId,
+			maybe_check_manager: Option<&AccountIdOf<T>>,
+			force_skip_refund: bool,
+			max_currencies: u32,
+		) -> Result<u32, DispatchError> {
+			let mut pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(
+				pool_details.bonded_currencies.len() <= max_currencies.saturated_into(),
+				Error::<T>::TooManyCurrencies
+			);
+
+			ensure!(
+				pool_details.state.is_live() || pool_details.state.is_refunding(),
+				Error::<T>::PoolNotLive
+			);
+
+			if let Some(caller) = maybe_check_manager {
+				ensure!(pool_details.is_manager(caller), Error::<T>::Unauthorized); // TODO: should this be permissionless if the pool is in refunding state?
```

Comment:
I would say yes. The tx should be permissionless

Reply:
I'm still undecided. I don't have any bigger concerns against it, besides that it may be a little ... unexpected? We could just keep the TODO around and decide on a final solution further down the road - the change would be minimal anyway